crates/roc_host/src/json_web_token.rs

```diff
@@ -0,0 +1,170 @@
+        self.name.inc();
+        self.value.inc();
+    }
+    fn dec(&mut self) {
+        self.name.dec();
+        self.value.dec();
+    }
+    fn is_refcounted() -> bool {
+        true
+    }
+}
+
+fn jwt_claims_to_roc(claims: BTreeMap<String, serde_json::Value>) -> RocList<JWTToRoc> {
+    let mut roc_claims = RocList::with_capacity(claims.len());
+
+    for (name, json_value) in claims {
+        roc_claims.push(JWTToRoc {
+            name: name.as_str().into(),
+            value: format!("{}", json_value).as_str().into(),
+        });
+    }
+
+    roc_claims
+}
+
+fn jwt_error_to_roc(err: jwt::error::Error) -> RocStr {
+    match err {
+        jwt::error::Error::AlgorithmMismatch(..) => RocStr::from("AlgorithmMismatch"),
+        jwt::error::Error::InvalidSignature => RocStr::from("InvalidSignature"),
+        jwt::error::Error::Format => RocStr::from("BadFormat"),
+        jwt::error::Error::Json(..) => RocStr::from("BadJson"),
+        jwt::error::Error::NoClaimsComponent => RocStr::from("MissingClaims"),
+        jwt::error::Error::NoHeaderComponent => RocStr::from("MissingHeader"),
+        jwt::error::Error::NoKeyId => RocStr::from("MissingKeyId"),
+        _ => RocStr::from(format!("{err}").to_string().as_str()),
+    }
+}
+
+pub fn jwt_verify(jwt: &JWTFromRoc) -> RocResult<RocList<JWTToRoc>, RocStr> {
+    use jwt::VerifyWithKey;
+
+    let token_str = jwt.token.as_str().to_string();
+
+    match jwt.algorithm() {
+        AlgorithmType::Hs256 => {
+            let maybe_key: Result<Hmac<Sha256>, RocStr> =
+                Hmac::new_from_slice(&jwt.secret.as_bytes())
+                    .map_err(|err| format!("{err}").as_str().into());
+
+            match maybe_key {
+                Ok(key) => {
+                    let claims: Result<BTreeMap<String, serde_json::Value>, jwt::error::Error> =
+                        token_str.verify_with_key(&key);
+
+                    match claims {
+                        Ok(claims) => RocResult::ok(jwt_claims_to_roc(claims)),
+                        Err(err) => RocResult::err(jwt_error_to_roc(err)),
+                    }
+                }
+                Err(err) => return RocResult::err(err),
+            }
+        }
+        AlgorithmType::Hs384 => {
+            let maybe_key: Result<Hmac<Sha384>, RocStr> =
+                Hmac::new_from_slice(&jwt.secret.as_bytes())
+                    .map_err(|err| format!("{err}").as_str().into());
+
+            match maybe_key {
+                Ok(key) => {
+                    let claims: Result<BTreeMap<String, serde_json::Value>, jwt::error::Error> =
+                        token_str.verify_with_key(&key);
+
+                    match claims {
+                        Ok(claims) => RocResult::ok(jwt_claims_to_roc(claims)),
+                        Err(err) => RocResult::err(jwt_error_to_roc(err)),
+                    }
+                }
+                Err(err) => return RocResult::err(err),
+            }
```

Comment:
Can we pull this repeated code out into a generic function. Just template it over `Hmac<Sha384>`? That would be a lot cleaner and make it so changes only have to be done in one place.

Reply:
I tried this, to refactor and make it dry, but I couldn't figure out how to make rust happy with the trait, and still ensure it's decoding the correct variant.

Reply:
Ah, I see.... The type Hmac takes is kinda insane:
```rust
where
    D: CoreProxy,
    D::Core: HashMarker
        + UpdateCore
        + FixedOutputCore
        + BufferKindUser<BufferKind = Eager>
        + Default
        + Clone,
    <D::Core as BlockSizeUser>::BlockSize: IsLess<U256>,
    Le<<D::Core as BlockSizeUser>::BlockSize, U256>: NonZero,
```

They really should have a wrapping trait so a user could do something like `T: HmacHash` and It would mean all of those things.

So to make a generic function just for hmac, I think it would be this:
```rust
pub fn jwt_verify<Sha>(
    token: &str,
    secret: &str,
) -> Result<HashMap<String, serde_json::Value>, jwt::Error>
where
    Sha: CoreProxy,
    Sha::Core: HashMarker
        + UpdateCore
        + FixedOutputCore
        + BufferKindUser<BufferKind = Eager>
        + Default
        + Clone,
    <Sha::Core as BlockSizeUser>::BlockSize: IsLess<U256>,
    Le<<Sha::Core as BlockSizeUser>::BlockSize, U256>: NonZero,
{
    let key: Hmac<Sha> = Hmac::new_from_slice(secret.as_bytes())?;
    let claims: HashMap<String, serde_json::Value> = token.verify_with_key(&key)?;
    Ok(claims)
}
```

so bad...